Library/OcCryptoLib/BigNumMontgomery.c

```diff
@@ -0,0 +1,677 @@
+  // variable value. For this reason, the first two iterations of the loop are
+  // unrolled. 2^Bits(Word) is represented as 0 as those two values are
+  // congruent modulo 2^Bits(Word), which is the variable storage space.
+  // The modulo operation is therefor implemented as a subtraction loop.
+  //
+
+  //
+  // === INITIALISATION ===
+  //
+  // Divisor  = 1U << sizeof(A->Words[0]) * 8;
+  // Dividend = A->Words[0];
+  //
+  // Y = 1;
+  // X = 0;
+  //
+  // === LOOP UNROLL 1) ===
+  //
+  // Div = Dividend / Divisor; // => 0
+  // Mod = Dividend % Divisor; // => A->Words[0]
+  //
+  // Dividend = Divisor;       // => 2^Bits(Word)
+  // Divisor  = Mod;           // => A->Words[0]
+  //
+  Dividend = 0;
+  Divisor  = A->Words[0];
+  //
+  // Tmp = Y - Div * X;        // => 1
+  // Y   = X;                  // => 0
+  // X   = Tmp;                // => 1
+  //
+  // === LOOP UNROLL 2) ===
+  //
+  // Div = Dividend / Divisor;
+  // Mod = Dividend % Divisor;
+  //
+  Div = 0;
+  do { 
+    Dividend -= Divisor;
+    ++Div;
+  } while (Dividend >= Divisor);
+
+  Mod = Dividend;
+  //
+  // Dividend = Divisor;
+  // Divisor  = Mod;
+  //
+  Dividend = Divisor;
+  Divisor  = Mod;
+  //
+  // Tmp = Y - Div * X;        // => -Div
+  // Y   = X;                  // => 1
+  // X   = Tmp;                // => -Div
+  //
+  Y = 1;
+  X = (OC_BN_WORD)0U - Div;
+
+  while (Divisor != 0) {
+    //
+    // FIXME: This needs a good source for an algorithm specification.
+    //
+    Div = Dividend / Divisor;
+    Mod = Dividend % Divisor;
+
+    Dividend = Divisor;
+    Divisor  = Mod;
+
+    Tmp = Y - Div * X;
+    Y   = X;
+    X   = Tmp;
+  }
+  //
+  // When the loop ends, Dividend contains the Greatest Common Divisor.
+  // If it is not 1, we cannot compute the Montgomery Inverse.
+  //
+  if (Dividend != 1) {
+    return 0;
+  }
+  //
+  // As per above, Y is 1 / A mod 2^#Bits(Word), so invert the result to yield
+  // -1 / A mod 2^#Bits(Word).
+  //
+  return (OC_BN_WORD)0U - Y;
+}
+
+/**
+  Calculates the Montgomery Inverse and R� mod N.
+
+  @param[in,out] RSqrMod  The buffer to return R� mod N into.
+  @param[in]     N        The Montgomery Modulus.
+
+  @returns  The Montgomery Inverse of N.
+
+**/
+OC_BN_WORD
+BigNumCalculateMontParams (
+  IN OUT OC_BN        *RSqrMod,
+  IN     CONST OC_BN  *N
+  )
+{
+  OC_BN_WORD N0Inv;
+  UINT32     NumBits;
+  UINTN      Size;
+  OC_BN      *RSqr;
+
+  ASSERT (RSqrMod != NULL);
+  ASSERT (N != NULL);
+  ASSERT (RSqrMod->NumWords == N->NumWords);
+  //
+  // Calculate the Montgomery Inverse.
+  // This is a reduced approach of the algorithmic -1 / N mod 2^#Bits(N),
+  // where the modulus is reduced from 2^#Bits(N) to 2^#Bits(Word). This
+  // reduces N to N[0] and yields an inverse valid in the Word domain.
+  //
+  N0Inv = BigNumMontInverse (N);
+  if (N0Inv == 0) {
+    return 0;
+  }
+
+  NumBits = BigNumSignificantBits (N);
+
+  OC_STATIC_ASSERT (
+    OC_BN_MAX_SIZE * 8 <= (MAX_UINTN / 2) - 2 - 7,
```

Comment:
Mhm, normally you would use CHAR_BIT constant instead of 8. We are not writing in plain C, but perhaps we could introduce at least some constant? These 8 and 7 all over the code make it quite a bit unintuitive.

Reply:
Agreed, but where would we put it? Do we want it in edk2?

Reply:
For now not, could go to OcMiscLib or OcGuardLib (makes some sense there). In future, EDK II.